nominal/core/datasource.py

```diff
@@ -0,0 +1,352 @@
+        end: str | datetime | IntegralNanosecondsUTC | None = None,
+        tags: dict[str, str] | None = None,
+    ) -> pd.DataFrame:
+        import pandas as pd
+
+        """Download a dataset to a pandas dataframe, optionally filtering for only specific channels of the dataset.
+
+        Args:
+        ----
+            channel_exact_match: Filter the returned channels to those whose names match all provided strings
+                (case insensitive).
+                For example, a channel named 'engine_turbine_rpm' would match against ['engine', 'turbine', 'rpm'],
+                whereas a channel named 'engine_turbine_flowrate' would not!
+            channel_fuzzy_search_text: Filters the returned channels to those whose names fuzzily match the provided
+                string.
+            tags: Dictionary of tags to filter channels by
+            start: The minimum data updated time to filter channels by
+            end: The maximum data start time to filter channels by
+
+        Returns:
+        -------
+            A pandas dataframe whose index is the timestamp of the data, and column names match those of the selected
+                channels.
+
+        Example:
+        -------
+        ```
+        import nominal as nm
+
+        rid = "..." # Taken from the UI or via the SDK
+        dataset = nm.get_dataset(rid)
+        df = dataset.to_pandas()
+        print(df.head())  # Show first few rows of data
+        ```
+
+        """
+        start_time = _SecondsNanos.from_flexible(start).to_api() if start else _MIN_TIMESTAMP.to_api()
+        end_time = _SecondsNanos.from_flexible(end).to_api() if end else _MAX_TIMESTAMP.to_api()
+        # Get all channels from the datasource
+        filtered_channels = self.search_channels(
+            exact_match=channel_exact_match,
+            fuzzy_search_text=channel_fuzzy_search_text,
+        )
+
+        batch_size = 20
+        all_dataframes = []
+
+        for channel_batch in batched(filtered_channels, batch_size):
+            export_request = self._construct_export_request(channel_batch, start_time, end_time, tags)
+            export_response = cast(
+                BinaryIO, self._clients.dataexport.export_channel_data(self._clients.auth_header, export_request)
+            )
+            batch_df = pd.DataFrame(pd.read_csv(export_response))
+            if not batch_df.empty:
+                all_dataframes.append(batch_df)
+
+        if not all_dataframes:
+            logger.warning(f"No data found for export from datasource {self.rid}")
+            raise RuntimeError(f"No data found for export from datasource {self.rid}")
+
+        result_df = pd.concat(all_dataframes, axis=0)
+        return result_df
+
+    def _construct_export_request(
+        self, channels: Sequence[Channel], start: api.Timestamp, end: api.Timestamp, tags: dict[str, str] | None
+    ) -> scout_dataexport_api.ExportDataRequest:
```

Comment:
can you move these sort of utility functions out of the class? these should really be free functions, the class doesn't need to be so long

Reply:
same with some of the new channel private methods